optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -58,55 +45,283 @@ that uses a functional approach:
+Penalties and rewards have several components:
+
+Constraint name::
+Human readable descriptive name for the constraint, which must be unique within the entire `ConstraintProvider`.
+Constraint weight::
+A constant score value indicating how much every breach of the constraint affects the score.
+Valid examples include `SimpleScore.ONE`, `HardSoftScore.ONE_HARD`, `HardMediumSoftScore.of(1, 2, 3)`, ...
+Constraint match weigher (Optional)::
+A function indicating how many times the constraint weight should be applied in the score.
+This number is typically determined from the contents of the constraint stream.
+Defaults to 1.
+Constraint package (Optional)::
+Java package that the constraint resides in.
+Defaults to the package where the `ConstraintProvider` resides.
+
+The ConstraintStreams API supports many different types of penalties and we suggest you browse the API in your IDE for
+the full list of method overloads. Here are some examples:
+
+Simple penalty::
+`penalize("Constraint name", SimpleScore.ONE)`
+Makes the score worse by 1 per every match in the constraint stream.
+In place of `SimpleScore.ONE`, you may use any `Score` depending on the needs of your problem.
+Dynamic penalty::
+`penalize("Constraint name", SimpleScore.ONE, Shift::getHours)`
+Makes the score worse by the number of hours in every matching `Shift` in the constraint stream.
+This is an example of using a constraint match weigher.
+Configurable penalty::
+`penalizeConfigurable("Constraint name")`
+Makes the score worse using constraint weights defined in <<constraintConfiguration,constraint configuration>>.
+Configurable dynamic penalty::
+`penalizeConfigurable("Constraint name", Shift::getHours)`
+Makes the score worse using constraint weights defined in <<constraintConfiguration,constraint configuration>>,
+multiplied by the number of hours in every matching `Shift` in the constraint stream.
+
+By replacing the keyword `penalize` by `reward` in the name of these building blocks, you will get operations that
+affect score in the opposite direction.
 
 
 [[constraintStreamsFilter]]
-== Filter
+=== Filter
+
+Filtering allows you to reduce the number of constraint matches in your stream.
+It first enumerate all constraint matches and then applies a predicate to filter some of them out.
+The predicate is a function that only returns `true` if the match is to continue in the stream.
+The following constraint stream will remove all of Beth's shifts from all `Shift` matches:
+
+[source,java,options="nowrap"]
+----
+    private Constraint penalizeAnnShifts(ConstraintFactory factory) {
+        return factory.from(Shift.class)
+                .filter(shift -> shift.getEmployeeName().equals("Ann"))
+                .penalize("Ann's shift", SimpleScore.ONE);
+    }
+----
+
+For a bi constraint match of `Shift` and `DayOff`, here's how to get a list of shifts where an employee has asked for a
```

Comment:
For -> from?

Also, I'd start this sentance with "For example, " (and without "here's how") so it's clear your switching from abstraction to example before any content.